ddtrace/internal/import_hooks/registry.py

```diff
@@ -0,0 +1,95 @@
+from collections import defaultdict
+import sys
+
+from ..logger import get_logger
+
+log = get_logger(__name__)
+
+
+class ModuleHookRegistry:
+    """
+    Registry to keep track of all module import hooks defined
+    """
+    __slots__ = ('hooks', )
+
+    def __init__(self):
+        """
+        Initialize a new registry
+        """
+        self.reset()
+
+    def register(self, name, func):
+        """
+        Register a new hook function for the provided module name
+
+        If the module is already loaded, then ``func`` is called immediately
+
+        :param name: The name of the module to add the hook for (e.g. 'requests', 'flask.app', etc)
+        :type name: str
+        :param func: The function to register as a hook
+        :type func: function(module)
+        """
+        self.hooks[name].add(func)
+
+        # Module is already loaded, call hook right away
+        if name in sys.modules:
+            func(sys.modules[name])
+
+    def deregister(self, name, func):
+        """
+        Deregister an already registered hook function
+
+        :param name: The name of the module the hook was for (e.g. 'requests', 'flask.app', etc)
+        :type name: str
+        :param func: The function that was registered previously
+        :type func: function(module)
+        """
+        # If no hooks exist for this module, return
```

Comment:
One of the advantage of the `defaultdict` is to avoid this kind of check. 🤷‍♂

Reply:
Yeah, but it means we'll end up populating a key for every module that is imported not every module that has a key.... although.... probably means should just be a `dict` with the manual checking instead of a `defaultdict`.

Reply:
moved to `dict` since I don't want to pollute `hooks` for every single module loaded, vs only modules which have hooks registered.